File: frontend/src/api/profiles.ts
import apiClient from './client';
import type { Profile, ProfileCreate, ProfileUpdate } from '../types';

export async function getProfiles(): Promise<Profile[]> {
  const res = await apiClient.get<Profile[]>('/profiles/');
  return res.data;
}

export async function getProfile(id: number): Promise<Profile> {
  const res = await apiClient.get<Profile>(`/profiles/${id}`);
  return res.data;
}

export async function createProfile(data: ProfileCreate): Promise<Profile> {
  const res = await apiClient.post<Profile>('/profiles/', data);
  return res.data;
}

export async function updateProfile(id: number, data: ProfileUpdate): Promise<Profile> {
  const res = await apiClient.put<Profile>(`/profiles/${id}`, data);
  return res.data;
}

File: backend/app/routers/profiles.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List

from app.database import get_db
from app.models.models import Profile
from app.schemas.schemas import ProfileCreate, ProfileRead, ProfileUpdate

router = APIRouter(prefix="/profiles", tags=["profiles"])


@router.get("/", response_model=List[ProfileRead])
def list_profiles(db: Session = Depends(get_db)):
    return db.query(Profile).all()


@router.get("/{profile_id}", response_model=ProfileRead)
def get_profile(profile_id: int, db: Session = Depends(get_db)):
    profile = db.query(Profile).filter(Profile.id == profile_id).first()
    if not profile:
        raise HTTPException(status_code=404, detail="Profile not found")
    return profile


@router.post("/", response_model=ProfileRead, status_code=201)
def create_profile(data: ProfileCreate, db: Session = Depends(get_db)):
    profile = Profile(**data.model_dump())
    db.add(profile)
    db.commit()
    db.refresh(profile)
    return profile


@router.put("/{profile_id}", response_model=ProfileRead)
def update_profile(profile_id: int, data: ProfileUpdate, db: Session = Depends(get_db)):
    profile = db.query(Profile).filter(Profile.id == profile_id).first()
    if not profile:
        raise HTTPException(status_code=404, detail="Profile not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(profile, field, value)
    db.commit()
    db.refresh(profile)
    return profile


@router.delete("/{profile_id}", status_code=204)
def delete_profile(profile_id: int, db: Session = Depends(get_db)):
    profile = db.query(Profile).filter(Profile.id == profile_id).first()
    if not profile:
        raise HTTPException(status_code=404, detail="Profile not found")
    db.delete(profile)
    db.commit()
